File: src/main/java/eu/siacs/conversations/xmpp/manager/EntityTimeManager.java
package eu.siacs.conversations.xmpp.manager;

import android.content.Context;
import android.util.Log;
import com.google.common.cache.CacheBuilder;
import com.google.common.cache.CacheLoader;
import com.google.common.cache.LoadingCache;
import com.google.common.collect.Collections2;
import com.google.common.collect.ImmutableSet;
import com.google.common.collect.Iterables;
import com.google.common.util.concurrent.Futures;
import com.google.common.util.concurrent.ListenableFuture;
import com.google.common.util.concurrent.MoreExecutors;
import eu.siacs.conversations.AppSettings;
import eu.siacs.conversations.Config;
import eu.siacs.conversations.xml.Namespace;
import eu.siacs.conversations.xmpp.Jid;
import eu.siacs.conversations.xmpp.XmppConnection;
import im.conversations.android.xmpp.model.error.Condition;
import im.conversations.android.xmpp.model.stanza.Iq;
import im.conversations.android.xmpp.model.time.Time;
import java.time.Duration;
import java.time.Instant;
import java.time.LocalTime;
import java.time.ZoneOffset;
import java.time.ZonedDateTime;
import java.util.Collection;
import java.util.List;
import java.util.Objects;
import java.util.Set;

public class EntityTimeManager extends AbstractManager {

    private final LoadingCache<Jid, ListenableFuture<EntityTime>> entityTimeCache =
            CacheBuilder.newBuilder()
                    .expireAfterWrite(Duration.ofHours(3))
                    .build(
                            new CacheLoader<>() {
                                @Override
                                public ListenableFuture<EntityTime> load(final Jid address) {
                                    return Futures.transform(
                                            zonedDateTime(address),
                                            zonedDateTime ->
                                                    new OffsetEntityTime(
                                                            Objects.requireNonNull(zonedDateTime)
                                                                    .getOffset()),
                                            MoreExecutors.directExecutor());
                                }
                            });
    private final AppSettings appSettings;

    public EntityTimeManager(final Context context, final XmppConnection connection) {
        super(context, connection);
        this.appSettings = new AppSettings(context);
    }

    public void request(final Iq request) {
        if (this.appSettings.isUseTor() || getAccount().isOnion()) {
            this.connection.sendErrorFor(request, new Condition.Forbidden());
            return;
        }
        final var from = request.getFrom();
        if (from == null) {
            this.connection.sendErrorFor(request, new Condition.NotAcceptable());
            return;
        }
        final var contact = getManager(RosterManager.class).getContactFromContactList(from);
        if (contact != null && contact.showInContactList() && this.appSettings.isEntityTime()) {
            Log.d(
                    Config.LOGTAG,
                    getAccount().getJid().asBareJid()
                            + ": responding to entity time request from "
                            + request.getFrom());
            this.connection.sendResultFor(request, new Time(ZonedDateTime.now()));
        } else {
            this.connection.sendErrorFor(request, new Condition.Forbidden());
        }
    }

    // TODO add a zoneDateTimeWithTimeout and use that in the cache

    public ListenableFuture<ZonedDateTime> zonedDateTime(final Jid address) {
        Log.d(Config.LOGTAG, "requesting entity time: " + address);
        final var future = this.connection.sendIqPacket(new Iq(Iq.Type.GET, address, new Time()));
        return Futures.transform(
                future,
                iq -> {
                    final var time = Objects.requireNonNull(iq).getOnlyExtension(Time.class);
                    if (time == null) {
                        throw new IllegalArgumentException("No valid time extension in response");
                    }
                    final var zonedDateTime = time.asZonedDateTime();
                    if (zonedDateTime == null) {
                        throw new IllegalArgumentException("No valid zoned date time in response");
                    }
                    return zonedDateTime;
                },
                MoreExecutors.directExecutor());
    }

    private ListenableFuture<List<EntityTime>> getEntityTimes(final Jid address) {
        final var presences = getManager(PresenceManager.class).getPresences(address);
        final Collection<ListenableFuture<EntityTime>> futures =
                Collections2.transform(
                        presences,
                        p -> {
                            final var fullAddress = Objects.requireNonNull(p).getFrom();
                            final var infoQuery = getManager(DiscoManager.class).get(fullAddress);
                            if (infoQuery == null || !infoQuery.hasFeature(Namespace.TIME)) {
                                return Futures.immediateFuture(new NoEntityTime());
                            }
                            final var entityTimeFuture = entityTimeCache.getUnchecked(fullAddress);
                            return Futures.catching(
                                    entityTimeFuture,
                                    Exception.class,
                                    ex -> new InvalidEntityTime(),
                                    MoreExecutors.directExecutor());
                        });
        return Futures.allAsList(futures);
    }

    public ListenableFuture<ZonedDateTime> getZonedDateTime(final Jid address) {
        if (!this.appSettings.isEntityTime()) {
            return Futures.immediateFailedFuture(
                    new IllegalStateException("Requesting entity time is disabled"));
        }
        return Futures.transform(
                getEntityTimes(address),
                entityTimes -> {
                    if (entityTimes == null) {
                        throw new IllegalStateException("No entity times available");
                    }
                    final var unknownAsNulls =
                            Collections2.transform(
                                    entityTimes,
                                    entityTime -> {
                                        if (entityTime
                                                instanceof
                                                OffsetEntityTime(ZoneOffset zoneOffset)) {
                                            return zoneOffset;
                                        } else {
                                            return null;
                                        }
                                    });
                    final Set<ZoneOffset> zoneOffsets =
                            ImmutableSet.copyOf(
                                    Collections2.filter(unknownAsNulls, Objects::nonNull));
                    if (zoneOffsets.size() == 1) {
                        return Instant.now().atZone(Iterables.getOnlyElement(zoneOffsets));
                    } else {
                        throw new IllegalStateException("Ambiguous time zones");
                    }
                },
                MoreExecutors.directExecutor());
    }

    public static boolean isNightTime(final ZonedDateTime zonedDateTime) {
        final var localTime = zonedDateTime.toLocalTime();
        return localTime.isAfter(LocalTime.of(22, 0)) || localTime.isBefore(LocalTime.of(8, 0));
    }

    public static boolean isDifferentTimeZone(final ZonedDateTime zonedDateTime) {
        final var local = ZonedDateTime.now();
        return !local.getOffset().equals(zonedDateTime.getOffset());
    }

    public sealed interface EntityTime {}

    public record NoEntityTime() implements EntityTime {}

    public record OffsetEntityTime(ZoneOffset zoneOffset) implements EntityTime {}

    public record InvalidEntityTime() implements EntityTime {}
}

File: src/main/java/eu/siacs/conversations/xmpp/manager/EntityTimeManager.kt
package eu.siacs.conversations.xmpp.manager

import android.util.Log
import com.google.common.cache.CacheBuilder
import com.google.common.cache.CacheLoader
import com.google.common.cache.LoadingCache
import com.google.common.collect.Collections2
import com.google.common.collect.ImmutableSet
import com.google.common.collect.Iterables
import com.google.common.util.concurrent.Futures
import com.google.common.util.concurrent.ListenableFuture
import com.google.common.util.concurrent.MoreExecutors
import eu.siacs.conversations.AppSettings
import eu.siacs.conversations.Config
import eu.siacs.conversations.entities.Conversation
import eu.siacs.conversations.services.XmppConnectionService
import eu.siacs.conversations.xml.Namespace
import eu.siacs.conversations.xmpp.Jid
import eu.siacs.conversations.xmpp.XmppConnection
import im.conversations.android.xmpp.model.error.Condition
import im.conversations.android.xmpp.model.stanza.Iq
import im.conversations.android.xmpp.model.time.Time
import java.time.Duration
import java.time.Instant
import java.time.LocalTime
import java.time.ZoneOffset
import java.time.ZonedDateTime
import java.util.Objects

class EntityTimeManager(private val service: XmppConnectionService, connection: XmppConnection) :
    AbstractManager(service.applicationContext, connection) {

    private val appSettings = AppSettings(context.applicationContext)

    private val entityTimeLoader: CacheLoader<Jid, ListenableFuture<EntityTime>> =
        object : CacheLoader<Jid, ListenableFuture<EntityTime>>() {
            override fun load(address: Jid): ListenableFuture<EntityTime> {
                return Futures.transform(
                    zonedDateTime(address),
                    { zonedDateTime ->
                        OffsetEntityTime(Objects.requireNonNull(zonedDateTime).offset)
                    },
                    MoreExecutors.directExecutor()
                )
            }
        }

    private val entityTimeCache: LoadingCache<Jid, ListenableFuture<EntityTime>> =
        CacheBuilder.newBuilder()
            .expireAfterWrite(Duration.ofHours(3))
            .build(entityTimeLoader)

    fun request(request: Iq) {
        if (this.appSettings.isUseTor() || getAccount().isOnion()) {
            this.connection.sendErrorFor(request, Condition.Forbidden())
            return
        }
        val from = request.getFrom()
        if (from == null) {
            this.connection.sendErrorFor(request, Condition.NotAcceptable())
            return
        }
        val contact = getManager(RosterManager::class.java).getContactFromContactList(from)
        if (contact != null && contact.showInContactList() && this.appSettings.isEntityTime()) {
            Log.d(
                Config.LOGTAG,
                "${getAccount().getJid().asBareJid()}: responding to entity time request from ${request.getFrom()}"
            )
            this.connection.sendResultFor(request, Time(ZonedDateTime.now()))
        } else {
            this.connection.sendErrorFor(request, Condition.Forbidden())
        }
    }

    fun zonedDateTime(address: Jid): ListenableFuture<ZonedDateTime> {
        Log.d(Config.LOGTAG, "requesting entity time: $address")
        val iqFuture = this.connection.sendIqPacket(Iq(Iq.Type.GET, address, Time()))
        val zonedEntityTimeFuture = Futures.transform(
            iqFuture,
            { iq ->
                val time = Objects.requireNonNull(iq).getOnlyExtension(Time::class.java)
                    ?: throw IllegalArgumentException("No valid time extension in response")
                val zonedDateTime = time.asZonedDateTime()
                    ?: throw IllegalArgumentException("No valid zoned date time in response")
                zonedDateTime
            },
            MoreExecutors.directExecutor()
        )
        Futures.addCallback(
            zonedEntityTimeFuture,
            object : com.google.common.util.concurrent.FutureCallback<ZonedDateTime> {
                override fun onSuccess(result: ZonedDateTime) {
                    service.updateConversationUi()
                }

                override fun onFailure(t: Throwable) {
                    Log.d(Config.LOGTAG, "could not request entity time from$address", t)
                }
            },
            MoreExecutors.directExecutor()
        )
        return zonedEntityTimeFuture
    }

    private fun getEntityTimes(address: Jid): ListenableFuture<List<EntityTime>> {
        val presences = getManager(PresenceManager::class.java).getPresences(address)
        val futures = Collections2.transform(presences) { p ->
            val fullAddress = Objects.requireNonNull(p).getFrom()
            val infoQuery = getManager(DiscoManager::class.java).get(fullAddress)
            if (infoQuery == null || !infoQuery.hasFeature(Namespace.TIME)) {
                return@transform Futures.immediateFuture<EntityTime>(NoEntityTime())
            }
            val entityTimeFuture = entityTimeCache.getUnchecked(fullAddress)
            val entityTimeTimeoutFuture = Futures.withTimeout(
                entityTimeFuture,
                Duration.ofSeconds(10L),
                FUTURE_TIMEOUT_EXECUTOR
            )
            Futures.catching(
                entityTimeTimeoutFuture,
                Exception::class.java,
                { InvalidEntityTime() },
                MoreExecutors.directExecutor()
            )
        }
        return Futures.allAsList(futures)
    }

    fun getZonedDateTime(address: Jid): ListenableFuture<ZonedDateTime> {
        if (!this.appSettings.isEntityTime()) {
            return Futures.immediateFailedFuture(
                IllegalStateException("Requesting entity time is disabled")
            )
        }
        val contact = getManager(RosterManager::class.java).getContactFromContactList(address)
            ?: return Futures.immediateFailedFuture(
                IllegalStateException("Requesting entity time is limited to contacts")
            )
        return Futures.transform(
            getEntityTimes(address),
            { entityTimes ->
                if (entityTimes == null) {
                    throw IllegalStateException("No entity times available")
                }
                val unknownAsNulls = Collections2.transform(entityTimes) { entityTime ->
                    if (entityTime is OffsetEntityTime) entityTime.zoneOffset else null
                }
                val zoneOffsets: Set<ZoneOffset> =
                    ImmutableSet.copyOf(unknownAsNulls.filterNotNull())
                if (zoneOffsets.size == 1) {
                    Instant.now().atZone(Iterables.getOnlyElement(zoneOffsets))
                } else {
                    throw IllegalStateException("Ambiguous time zones")
                }
            },
            MoreExecutors.directExecutor()
        )
    }

    sealed interface EntityTime

    class NoEntityTime : EntityTime

    class OffsetEntityTime(val zoneOffset: ZoneOffset) : EntityTime

    class InvalidEntityTime : EntityTime

    companion object {
        @JvmStatic
        fun isNightTime(zonedDateTime: ZonedDateTime): Boolean {
            val localTime = zonedDateTime.toLocalTime()
            return localTime.isAfter(LocalTime.of(22, 0)) || localTime.isBefore(LocalTime.of(8, 0))
        }

        @JvmStatic
        fun isDifferentTimeZone(zonedDateTime: ZonedDateTime): Boolean {
            val local = ZonedDateTime.now()
            return local.offset != zonedDateTime.offset
        }

        @JvmStatic
        fun noRecentMessages(conversation: Conversation): Boolean {
            val duration = Duration.between(conversation.getLastReceived(), Instant.now())
            return DURATION_RECENT_MESSAGE.compareTo(duration) < 0
        }

        private val DURATION_RECENT_MESSAGE: Duration = Duration.ofMinutes(12)
    }
}
